graph-builder/src/metadata.rs

```diff
@@ -0,0 +1,107 @@
+//! This module implements the fetching of dynamic metadata
+
+use failure::{Error, Fallible};
+use futures::future::Future;
+use quay;
+use registry;
+use std::path::PathBuf;
+
+pub static DEFAULT_QUAY_LABEL_FILTER: &str = "com.openshift.upgrades.graph";
+pub static MANIFESTREF_KEY: &str = "com.openshift.upgrades.graph.release.manifestref";
+
+pub struct MetadataFetchArgs {
+    manifestref: String,
+}
+
+impl MetadataFetchArgs {
+    pub fn new(manifestref: String) -> Self {
+        MetadataFetchArgs { manifestref }
+    }
+}
+
+pub trait AsyncMetadataFetcher<T, E> {
+    fn fetch_metadata(&self, fetch_args: &MetadataFetchArgs) -> Box<Future<Item = T, Error = E>>;
+}
+
+/// Convenience type for a metadata entry
+type MetadataEntry = (String, String);
+
+/// Convenience type for fetch_args()'s metadata_fetcher argument
+pub type MetadataFetcher<'a> = Box<AsyncMetadataFetcher<Vec<MetadataEntry>, Error> + 'a>;
+
+pub struct QuayMetadataFetcher<'a> {
+    client: quay::v1::Client,
+    label_filter: &'a str,
+    repo: &'a str,
+}
+impl<'a> AsyncMetadataFetcher<Vec<MetadataEntry>, Error> for QuayMetadataFetcher<'a> {
+    fn fetch_metadata(
+        &self,
+        fetch_args: &MetadataFetchArgs,
+    ) -> Box<Future<Item = Vec<MetadataEntry>, Error = Error>> {
+        Box::new(
+            self.client
+                .get_labels(
+                    self.repo.to_string(),
+                    fetch_args.manifestref.to_string(),
+                    Some(self.label_filter.to_string()),
```

Comment:
It looks like you had to wire lifetimes through everywhere, but eventually the future needs to take ownership of these three parameters, thus cloning them here. It may be worth just taking ownership (in function inputs and in struct fields) and break the chain of lifetimes.

Up for your consideration though, I don't know which approach makes more sense. I just noticed it when following the lifetimes through.

Reply:
Indeed it doesn't make sense to go through the lifetime declarations and ending up cloning it in the end, thanks for pointing this one out :smile: